### PR TITLE
feat(core): Handle nodes with multiple inputs and connections during partial executions

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

This PR re-implements the functionality for recreating the execution stack from run data, start nodes and pinned data.

The current implementation was copied from the old partial execution logic and didn't account for nodes with multiple inputs, but not all of them having data.

Here is an example of how it was working before:

https://github.com/user-attachments/assets/797d8288-261c-4dd9-a4d6-bef37250a17e

The first execution is a full execution. The merge node runs twice and produces 3 items. One run with 2 items and one run with 1 item.

The other 2 executions execute node1 and node2 before merge, respectively and produce different results, neither of them matching what the full execution produces.

This is how it works now:

https://github.com/user-attachments/assets/13aa8c18-86bb-490c-a2bd-dda817481baf

All 3 executions leave the `merge` node being executed twice and producing 3 items.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] ~[Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.~
- [x] Tests included. 
- [ ] ~PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)~

### CODE REVIEW

File: packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts [hunk 1]
```diff
@@ -62,17 +73,8 @@ describe('recreateNodeExecutionStack', () => {
...
-			node: {
-				'0': {
-					main: [
-						{ previousNode: 'trigger', previousNodeOutput: undefined, previousNodeRun: undefined },
-					],
-				},
-			},
-		});
+		expect(waitingExecution).toEqual({});
+		expect(waitingExecutionSource).toEqual({});
```
Comment (despairblue) [inline packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts:77]: node is already on the execution stack, it should not be in the waiting stack

File: packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts [hunk 2]
```diff
@@ -105,8 +107,8 @@ describe('recreateNodeExecutionStack', () => {
 			},
 		]);
 
-		expect(waitingExecution).toEqual({ node: { '0': { main: [null] } } });
-		expect(waitingExecutionSource).toEqual({ node: { '0': { main: [null] } } });
+		expect(waitingExecution).toEqual({});
+		expect(waitingExecutionSource).toEqual({});
```
Comment (despairblue) [inline packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts:111]: We don't have to add node to the waiting stack. The execution engine will do that after execution trigger.

File: packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts [hunk 3]
```diff
@@ -251,19 +252,8 @@ describe('recreateNodeExecutionStack', () => {
...
-				'0': {
-					main: [
-						{ previousNode: 'node1', previousNodeOutput: undefined, previousNodeRun: undefined },
-						{ previousNode: 'node2', previousNodeOutput: undefined, previousNodeRun: undefined },
-					],
-				},
-			},
-		});
+		expect(waitingExecution).toEqual({});
+		expect(waitingExecutionSource).toEqual({});
```
Comment (despairblue) [inline packages/core/src/PartialExecutionUtils/__tests__/recreateNodeExecutionStack.test.ts:256]: node3 is twice on the exec stack, no need to have it on the waiting stack too

File: packages/core/src/PartialExecutionUtils/getIncomingData.ts [hunk 4]
```diff
@@ -20,3 +20,26 @@ export function getIncomingData(
...
+
+	for (let runIndex = 0; runIndex < maxRunIndexes; runIndex++) {
+		const data = getIncomingData(runData, nodeName, runIndex, connectionType, outputIndex);
+
+		if (data && data.length > 0) {
+			return { data, runIndex };
+		}
+	}
+
+	return;
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getIncomingData.ts:44]:
```suggestion
	return undefined;
```

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts [hunk 5]
```diff
@@ -13,6 +13,18 @@ function sortByInputIndexThenByName(
 	}
 }
 
+type Group = {
+	complete: boolean;
+	connections: GraphConnection[];
+};
+
+function newGroup() {
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:21]:
```suggestion
function newGroup(): Group {
```

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts [hunk 6]
```diff
@@ -58,55 +70,79 @@ function sortByInputIndexThenByName(
...
 		}
 	}
 
 	sortedConnectionsWithData.sort(sortByInputIndexThenByName);
+	sortedConnectionsWithoutData.sort(sortByInputIndexThenByName);
 
-	const groups: GraphConnection[][] = [];
-	let currentGroup: GraphConnection[] = [];
+	const groups: Group[] = [];
+	let currentGroup: Group = newGroup();
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:104]: After we add return type to `newGroup` the explicit type is not needed anymore

```suggestion
	let currentGroup = newGroup();
```

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts [hunk 7]
```diff
@@ -58,55 +70,79 @@ function sortByInputIndexThenByName(
...
-		);
-		const connection: GraphConnection | undefined =
-			sortedConnectionsWithData[connectionWithDataIndex];
-
-		if (connection === undefined) {
-			groups.push(currentGroup);
-			currentGroup = [];
-			currentInputIndex = -1;
-			continue;
+	while (sortedConnectionsWithData.length > 0 || sortedConnectionsWithoutData.length > 0) {
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:107]: Is there a risk that this might cause an infinite loop in some obscure scenario?

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts [hunk 8]
```diff
@@ -13,6 +13,18 @@ function sortByInputIndexThenByName(
 	}
 }
 
+type Group = {
+	complete: boolean;
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:17]: Could document what complete means here

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts [hunk 9]
```diff
@@ -13,6 +13,18 @@ function sortByInputIndexThenByName(
 	}
 }
 
+type Group = {
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:16]: What group is this? Could we be more explicit? Is it for example a `SourceConnectionGroup` or `IncomingConnectionGroup`?

File: packages/core/src/PartialExecutionUtils/recreateNodeExecutionStack.ts [hunk 10]
```diff
@@ -84,89 +118,94 @@ export function recreateNodeExecutionStack(
...
+					for (const incomingConnection of sourceData.connections) {
+						let runIndex = 0;
+						const sourceNode = incomingConnection.from;
+
+						if (pinData[sourceNode.name]) {
+							incomingData.push(pinData[sourceNode.name]);
+						} else {
+							a.ok(
+								runData[sourceNode.name],
+								`Start node(${incomingConnection.to.name}) has an incoming connection with no run or pinned data. This is not supported. The connection in question is "${sourceNode.name}->${startNode.name}". Are you sure the start nodes come from the "findStartNodes" function?`,
```
Comment (tomi) [inline packages/core/src/PartialExecutionUtils/recreateNodeExecutionStack.ts:137]: Nice 👍

**Review by tomi — COMMENTED** [review]
Very comprehensive unit tests 👏 Few comments mainly about naming and documentation as well as couple questions

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts (hunk 7 above)
Comment (despairblue) [reply packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:107]: Yeah, that's hard to follow. 🤔 

Technically it should always terminate unless it contains a connection with a negative `inputIndex`.

I could iterate over all connections, find the lowest and highest `inputIndex`es and use them inside a for loop 🤔

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts (hunk 9 above)
Comment (despairblue) [reply packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts]: I renamed it ❤️

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts (hunk 8 above)
Comment (despairblue) [reply packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:25]: added docs

File: packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts (hunk 7 above)
Comment (despairblue) [reply packages/core/src/PartialExecutionUtils/getSourceDataGroups.ts:122]: I tried rewriting it. It just turns messier than this. Let me know if you have an idea, otherwise I think the tests are pretty complete now.

**Review by tomi — APPROVED** [review]
🚀

**Comment by janober** [comment]
Got released with [`n8n@1.66.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.66.0)
